### DISCUSSION TITLE
How to render MDX string as a react component?

### DISCUSSION THREAD

**Discussion by melvin2016** [vercel/next.js#13901] [Help] [answered]
This is my `pages` directory.

<img width="391" alt="Screenshot 2020-06-08 at 8 35 20 PM" src="https://user-images.githubusercontent.com/22199666/84046396-adf01580-a9c7-11ea-9bd2-3e6e0f2cab58.png">

This is my `posts` directory with the .mdx files

<img width="391" alt="Screenshot 2020-06-08 at 8 41 16 PM" src="https://user-images.githubusercontent.com/22199666/84047160-99f8e380-a9c8-11ea-864c-c92754dd687b.png">

I have a hierarchy like this for my blog. I'm dynamically creating blog links using `getStaticPaths` function like this.

```javascript
export async function getStaticPaths() {
  const fs = require("fs");
  const path = require("path");
  const filesString = fs.readdirSync(path.join("content", "posts/")).toString();
  const linksArray = filesString.split(",");
  const params = linksArray.map((link) => ({
    params: {
      link: link.replace(".mdx", ""),
    },
  }));

  return {
    paths: params,
    fallback: false,
  };
}
```

And providing data to each link using the `getStaticProps` function

```javascript
export async function getStaticProps({ params }) {
  const mdx = require("@mdx-js/mdx");
  const fs = require("fs");
  const path = require("path");
  const matter = require("gray-matter");
  // Get File and get YAML data from markdown
  const post = fs
    .readFileSync(path.join("content", "posts/", `${params.link}.mdx`))
    .toString();
  const { data, content } = matter(post);

  // Serialize date
  const dateOptions = { year: "numeric", month: "long", day: "numeric" };
  const formattedDate = data.date.toLocaleDateString("en-US", dateOptions);

  // Make metadata
  const frontMatter = {
    ...data,
    date: formattedDate,
    link: params.link,
  };

  const jsx = mdx.sync(content);
  console.log(jsx);

  return {
    props: {
      content: `# ${data.title}
      \n<p className="date">${formattedDate}<p>
      \n${content}`,
      frontMatter,
    },
  };
}
```

 But I want to render my JSX which is generated from the `mdx.sync()` function. How can I do this?

**Comment by Vadorequest** [comment]
I'm also trying to achieve something very similar.

- https://github.com/vercel/next.js/discussions/13843 I've made a POC with Next a few days ago (just discovered MDX), but it doesn't support what you want to do, only from file.
- https://github.com/mdx-js/mdx/issues/126

The official MDX doc seem to use a **virtual file** (need to dig in, may be the solution):
https://mdxjs.com/advanced/api#compiler

I don't know if that helps, hope so!

**Comment by mixja** [comment, marked as answer]
See https://github.com/hashicorp/next-mdx-remote

**Reply by melvin2016** [reply]
Just what I was looking for I think. +1

**Comment by deadcoder0904** [comment]
@melvin2016 `next-mdx-remote` works like a charm but if you want to put images in the same folder as the posts it doesn't work.

I have the same structure as you & this worked like a charm for me → https://github.com/vercel/next.js/issues/9524#issuecomment-580239600

I have a minimal repo if you want to have a look (images aren't working currently but I'm working on it) → https://github.com/deadcoder0904/blog-mdx-next/
